feat(py/trtorch/ptq): Implement INT8 Python API for PTQ

peri044 opened the pull request:
# Description

Implement INT8 PTQ Python API  support. Discussion thread : [https://github.com/NVIDIA/TRTorch/discussions/346](url)
Three main goals addressed:
- Implement data loader calibrators and cache variants in python
- Re-export TensorRT bindings through TRTorch. This includes support for using TRT calibrators directly in TRTorch without symbol conflicts
- Use python meta classes instead of writing separate data loader calibrators for different types of calibration algorithm

Fixes https://github.com/NVIDIA/TRTorch/issues/173, https://github.com/NVIDIA/TRTorch/issues/57

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

narendasan commented on cpp/api/include/trtorch/trtorch.h:
```diff
@@ -34,7 +34,7 @@ class IInt8Calibrator;
 #endif // DOXYGEN_SHOULD_SKIP_THIS
 
-#include "trtorch/macros.h"
+#include "trtorch/macros.h" // 
```
Seems like a typo

peri044 replied:
Fixed it.

narendasan commented on py/trtorch/__init__.py:
```diff
@@ -9,6 +9,7 @@
 
 from trtorch._version import __version__
 from trtorch._compiler import *
+from trtorch.ptq import *
```
Does this dump everything in the `trtorch` namespace? I would like to see the ptq stuff in its own subpackage

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -9,12 +9,106 @@
+  bool getBatch(void* bindings[], const char* names[], int nbBindings) noexcept override {
+    py::gil_scoped_acquire gil{};
+
+    py::function pyGetBatch = tensorrt::utils::getOverload(static_cast<Derived*>(this), "get_batch");
```
What is the tensorrt namespace?

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -9,12 +9,106 @@
+    py::function pyWriteCalibrationCache =
+        tensorrt::utils::getOverload(static_cast<Derived*>(this), "write_calibration_cache");
+
+#if PYBIND11_VERSION_MAJOR < 2 || PYBIND11_VERSION_MAJOR == 2 && PYBIND11_VERSION_MINOR < 6
```
We dont need to worry about old pybind versions I think

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up  to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -9,12 +9,106 @@
+  }
+};
+
+class pyIInt8LegacyCalibrator : public pyCalibratorTrampoline<IInt8LegacyCalibrator> {
```
Do we need to support this at all? Since I think for the DataLoader calibrator we should only support current features. Is there a deprecation plan for the Legacy Calibrator?

peri044 replied:
I checked with TRT and this will atleast stay for the next release. I haven't seen any deprecation plan yet. I think there's no harm in supporting this. This still works when I ran the calibration.

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -102,10 +196,46 @@ PYBIND11_MODULE(_C, m) {
       .value("default", EngineCapability::kDEFAULT, "Use default behavior");
 
+  py::enum_<CalibrationAlgoType>(m, "CalibrationAlgo", py::module_local(), "Type of calibration algorithm")
+      .value("LEGACY_CALIBRATION", CalibrationAlgoType::kLEGACY_CALIBRATION)
```
same question about legacy support

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -102,10 +196,46 @@ PYBIND11_MODULE(_C, m) {
+      .def("get_batch_size", &IInt8Calibrator::getBatchSize, "Get batch size")
+      .def("get_algorithm", &IInt8Calibrator::getAlgorithm, "Get algorithm");
+
+  py::class_<IInt8LegacyCalibrator, IInt8Calibrator, pyIInt8LegacyCalibrator>(
```
Same again

narendasan commented on py/trtorch/csrc/utils.h:
```diff
@@ -0,0 +1,138 @@
+#include <string>
+#include "NvInfer.h"
+
+namespace tensorrt {
```
put this in the `trtorch::pyapi::util` namespace. Also lets call it the `util` namespace instead of utils to be consistent with core

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up  to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/utils.h:
```diff
@@ -0,0 +1,138 @@
+    case nvinfer1::DataType::kBOOL:
+      return 1;
+  }
+  return -1;
```
We should just throw an error here and return 0 to suppress any compiler warnings. Returning -1 for an unsigned type is probably dangerous

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up  to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/utils.h:
```diff
@@ -0,0 +1,138 @@
+  } else if (type.is(py::dtype("b1"))) {
+    return nvinfer1::DataType::kBOOL;
+  }
+  std::cout << "[ERROR] Unsupported numpy data type: " << type.kind() << type.itemsize() * 8
```
Use TRTorch errors here

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up  to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/utils.h:
```diff
@@ -0,0 +1,138 @@
+  return py::array{nptype(self.type), self.count, self.values, py::cast(self)};
+};
+
+inline size_t volume(const nvinfer1::Dims& dims) {
```
Core should already have this

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up  to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/utils.h:
```diff
@@ -0,0 +1,138 @@
+py::function getOverload(const T* self, const std::string& overloadName) {
+  py::function overload = py::get_override(self, overloadName.c_str());
+  if (!overload) {
+    std::cerr << "Method: " << overloadName << " was not overriden. Please provide an implementation for this method.";
```
TRTorch errors

peri044 replied:
Fixed it. This was inherited from TRT source code. I cleaned it up  to be as minimal and relevant as possible.

narendasan commented on py/trtorch/csrc/utils.h:
```diff
@@ -0,0 +1,138 @@
+  return overload;
+}
+
+// Deprecation helpers
```
We should get rid of all this depreciation stuff

peri044 replied:
Fixed it. Cleaned it up.

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,61 @@
+from trtorch._compile_spec import _parse_compile_spec
+from trtorch._version import __version__
+from types import FunctionType
+import tensorrt as trt
```
We should not be importing tensorrt python api. Any thing we need from here should be reexported from our bindings.

peri044 replied:
Fixed it. Cleaned it up.

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,61 @@
+from trtorch._version import __version__
+from types import FunctionType
+import tensorrt as trt
+
```
We should be formulating this as a class `DataLoaderCalibrator` or something that at least looks like a class to the user.

narendasan replied:
So it should feel like you are using a constructor. 

User: 
```py

calibrator = trtorch.ptq.DataLoaderCalibrator(dl, 
                                                cache_file="<PATH>", 
                                                use_cache=True, 
                                                algo_type=trtorch.ptq.EntropyCalibrator2)

```

narendasan replied:
I think its better for it to look like the PyTorch Python API for DataLoaders which is formulated as a class. The `make_` syntax is based on how PyTorch does dataloaders in C++ and is consistent with things like shared pointers that arent in Python. I think function to function alignment with the C++ is less important than making things feel natural in their settings

peri044 replied:
Got it.

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,61 @@
+                       'write_calibration_cache' : write_calibration_cache}
+
+    # Using type metaclass to construct calibrator class based on algorithm type
+    if algo_type == trtorch._C.CalibrationAlgo.ENTROPY_CALIBRATION:
```
Can you reexport these in the ptq namespace as well for the user to define the calibration algo?

peri044 replied:
Done

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,61 @@
+    with open(self.cache_file, "wb") as f:
+        f.write(cache)
+
+def make_int8_calibrator(dataloader, cache_file, use_cache, algo_type):
```
Lets use kwargs for everything except the dataloader

peri044 replied:
Replaced this with a class

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,61 @@
+    print("Calibrating batch: ", self.current_batch_idx)
+    # Treat the first element as input and others as targets.
+    if isinstance(batch, list):
+        batch = batch[0].to(torch.device('cuda:0'))
```
This doesnt seem right. Does this work in multigpu? Im not sure what the best way to deal with this is.

narendasan replied:
It would be ideal if we could try to re-consume some of the user data from the CompileSpec but its probably going to be messy. I can remember if you can move all dataloader tensors at once at the beginning. I would be fine having the user set it properly.

peri044 replied:
This is written for my specific example which used off-the-shelf cifar10 dataloader code which returned (`inputs`, `targets`). Ideally, the dataloader is expected to be written by user where it is only supposed to return just `inputs`. 
I plan to remove this 
```
if isinstance(batch, list):
        batch = batch[0].to(torch.device('cuda:0'))
```
and just return batch.to(...).data_ptr(). What do you think ?

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,61 @@
+    with open(self.cache_file, "wb") as f:
+        f.write(cache)
+
+def make_int8_calibrator(dataloader, cache_file, use_cache, algo_type):
```
We should just call this `DataLoaderCalibrator` and pretend its a class constructor instead of replicating C++ style apis

peri044 replied:
Refactored.

narendasan commented:
Also seems like we still need the `CacheCalibrator` implementation in addition to tests for all three paths

peri044 commented:
@narendasan Cache calibrator works though the same code. When `use_cache=True`, we use `get_batch_cache_mode` to return `None` instead of `get_batch` which usually returns a batch. This was the only difference I found between non-cache and cache calibrators. 
Also, I cleaned up and addressed some comments. Posted my questions accordingly  for some of the questions.

narendasan commented:
Right but say I have a cache but no dataloader how do I use DataLoaderCalibrator?

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -29,7 +30,7 @@ class pyCalibratorTrampoline : public Derived {
     py::gil_scoped_acquire gil{};
 
-    py::function pyGetBatch = tensorrt::utils::getOverload(static_cast<Derived*>(this), "get_batch");
+    py::function pyGetBatch = trtorch::util::getOverload(static_cast<Derived*>(this), "get_batch");
```
This should be `trtorch::pyapi::util`

peri044 replied:
Done

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -44,7 +45,7 @@ class pyCalibratorTrampoline : public Derived {
 
     py::function pyReadCalibrationCache =
-        tensorrt::utils::getOverload(static_cast<Derived*>(this), "read_calibration_cache");
+        trtorch::util::getOverload(static_cast<Derived*>(this), "read_calibration_cache");
```
This should be trtorch::pyapi::util

peri044 replied:
Done

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -6,7 +6,11 @@
 from types import FunctionType
-import tensorrt as trt
+
+ENTROPY_CALIBRATION = trtorch._C.CalibrationAlgo.ENTROPY_CALIBRATION
```
Is it possible to group these in an enum or something called `CalibrationAlgo`? I think we do something similar for logging

peri044 replied:
Done. Is the new class okay or any better way to expose them ?

narendasan commented:
Also seems like we are missing tests

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -9,12 +9,95 @@
+#include "util.h"
 
+using namespace nvinfer1;
+using namespace pybind11::literals;
```
Just a style thing, I prefer to not use `using` since it masks namespaces later, if theres something we are using from TensorRT or pybind its easier if its clear where that object comes from.

narendasan replied:
so for literals it would be better to do `py::literals::OBJECT`

peri044 replied:
Changed it.

narendasan commented on py/trtorch/csrc/util.h:
```diff
@@ -0,0 +1,33 @@
+#include <functional>
+#include <iostream>
+#include <string>
+#include "core/util/logging/TRTorchLogger.h"
```
Just include "core/prelude.h" It has the macros too

peri044 replied:
Done.

narendasan commented on py/trtorch/ptq.py:
```diff
@@ -0,0 +1,106 @@
+    def __init__(self, cache_file, algo_type):
+        self.algo_type = algo_type
+        if os.path.isfile(cache_file):
+            print("Using cache file for calibration ", cache_file)
```
We can use the logger for this

peri044 replied:
Fixed

peri044 commented on py/trtorch/csrc/util.h:
```diff
@@ -0,0 +1,33 @@
+                    " was not overriden. Please provide an implementation for this method."};
+    // std::cerr << "Method: " << overloadName << " was not overriden. Please provide an implementation for this
+    // method.";
+    core::util::logging::get_logger().log(core::util::logging::LogLevel::kERROR, msg);
```
Done.

peri044 commented:
Working on test case and documentation. I will add it soon.

narendasan commented on docs/_sources/tutorials/ptq.rst.txt:
```diff
@@ -132,11 +135,76 @@ Then all thats required to setup the module for INT8 calibration is to set the f
+    compile_spec = {
+             "input_shapes": [[1, 3, 32, 32]],
+             "op_precision": torch.int8,
+             "calibrator": self.calibrator(),
```
This calling syntax is going to trip people up since people will probably assume that the calibrator is an object. Is there a way to do what we need to do without this syntax?

narendasan replied:
`__new__` might be a way to do this

peri044 replied:
Fixed it.

narendasan commented on tests/py/test_ptq_dataloader_calibrator.py:
```diff
@@ -0,0 +1,95 @@
+        with torch.no_grad():
+            idx = 0
+            for data, labels in testing_dataloader:
+                data, labels = data.cuda(), labels.cuda(async=True)
```
This uses a deprecated api, we should probably update it.

narendasan commented on tests/py/test_ptq_trt_calibrator.py:
```diff
@@ -0,0 +1,112 @@
+            'write_calibration_cache': write_calibration_cache
+        }
+
+        self.calibrator = type('DataLoaderCalibrator', (trt.IInt8EntropyCalibrator,), self.attribute_mapping)()
```
I get that this is probably easier but we should at least try to implement the calibrator like they do in the TRT docs even if it just wraps the dataloadercalibrator functions. Like actually create a class that inherits properly and all

narendasan commented on docs/_sources/tutorials/ptq.rst.txt:
```diff
@@ -132,11 +135,73 @@ Then all thats required to setup the module for INT8 calibration is to set the f
+
+.. code-block:: python
+
+    self.testing_dataset = torchvision.datasets.CIFAR10(root='./data',
```
lets remove the `self`s here

narendasan commented on py/trtorch/csrc/trtorch_py.cpp:
```diff
@@ -9,12 +9,97 @@
 #include "torch/torch.h"
+#include "util.h"
 
+// using namespace nvinfer1;
```
delete this line
